### PR TITLE
MINIFI-1354: Memory leaks in nanofi and its unittests

### PR DESCRIPTION
**Pull request by martinzink**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE file?
- [x] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nanofi/src/api/nanofi.cpp [hunk 1]
```diff
@@ -131,20 +131,26 @@ nifi_instance *create_instance_repo(const char *url, nifi_port *port, const char
...
-    nifi_port port;
-    auto port_str = utils::IdGenerator::getIdGenerator()->generate().to_string();
-    port.port_id = const_cast<char*>(port_str.c_str());
-    instance = create_instance("internal_standalone", &port);
-  }
+  if (instance == nullptr)
+    instance = acquire_standalone_instance();
   auto flow = create_new_flow(instance);
-  std::shared_ptr<ExecutionPlan> plan(flow);
+  std::shared_ptr<ExecutionPlan> plan(flow, free_flow);
```
Comment (martinzink) [inline nanofi/src/api/nanofi.cpp:153]: This caused the mismatched malloc vs operator delete (which has undefined behavior)

File: nanofi/src/api/nanofi.cpp [hunk 2]
```diff
@@ -131,20 +131,26 @@ nifi_instance *create_instance_repo(const char *url, nifi_port *port, const char
   return instance;
 }
 
+nifi_instance * acquire_standalone_instance () {
+  if (standalone_instance == nullptr) {
+    nifi_port port;
+    auto port_str = utils::IdGenerator::getIdGenerator()->generate().to_string();
+    port.port_id = const_cast<char *>(port_str.c_str());
```
Comment (lordgamez) [inline nanofi/src/api/nanofi.cpp:138]: We could initialize the port here like
```
nifi_port port{const_cast<char *>(port_str.c_str())};
```

**Review by lordgamez — COMMENTED** [review]
Looks good, thanks for the fix. My question is, that would it be possible to have an ASAN build in the CI and have the sanitizer checks there as well?

PS: The title should contain MINIFICPP instead of MINIFI :)

**Comment by martinzink** [comment]
> Looks good, thanks for the fix. My question is, that would it be possible to have an ASAN build in the CI and have the sanitizer checks there as well?

Yeah that would be ideal, unfortunetly there are still couple of issues we need to fix before that can be enabled. e.g. [MINIFICPP-839](https://issues.apache.org/jira/browse/MINIFICPP-839)

**Comment by lordgamez** [comment]
> > Looks good, thanks for the fix. My question is, that would it be possible to have an ASAN build in the CI and have the sanitizer checks there as well?
> 
> Yeah that would be ideal, unfortunetly there are still couple of issues we need to fix before that can be enabled. e.g. [MINIFICPP-839](https://issues.apache.org/jira/browse/MINIFICPP-839)

Oh, thanks for the info, I thought only these issues were found by the sanitizer.

File: nanofi/src/api/nanofi.cpp (hunk 2 above)
Comment (martinzink) [reply nanofi/src/api/nanofi.cpp]: changed in [6e43953](https://github.com/martinzink/nifi-minifi-cpp/commit/6e4395376e34445296946998d0110de92b24511e)

**Review by lordgamez — APPROVED** [review]

File: nanofi/tests/CLogAggregatorTests.cpp [hunk 3]
```diff
@@ -31,18 +31,19 @@
 
 #include "CTestsBase.h"
 
-void test_lists_equal(token_list * tknlist, const std::vector<std::string>& sv) {
+void test_lists_equal(const token_list * tknlist, const std::vector<std::string>& sv) {
     REQUIRE(tknlist != NULL);
     if (sv.empty()) {
```
Comment (adamdebreceni) [inline nanofi/tests/CLogAggregatorTests.cpp:36]:
```suggestion
    REQUIRE(tknlist->size == sv.size());
    token_node *node = tknlist->head;
    for (const auto& s : sv) {
      REQUIRE(node);
      REQUIRE(strcmp(s.c_str(), node->data) == 0);
      node = node->next;
    }
    REQUIRE(node == nullptr);
```

Comment (martinzink) [reply nanofi/tests/CLogAggregatorTests.cpp]: good idea, changed it in [e010b24](https://github.com/apache/nifi-minifi-cpp/pull/1035/commits/e010b24559ee441d528884f70e05eb68aec08400)

**Review by adamdebreceni — APPROVED** [review]

**Review by fgerlits — APPROVED** [review]
LGTM

**Pull request closed without merging**